core/src/main/scala/org/apache/spark/sql/streaming/DirectKafkaStreamSource.scala

```diff
@@ -45,34 +43,39 @@ final class DirectKafkaStreamRelation(
     @transient override val sqlContext: SQLContext,
     opts: Map[String, String],
     override val schema: StructType)
-    extends StreamBaseRelation(opts)
+  extends StreamBaseRelation(opts)
     with Logging with Serializable {
 
-  val topicsSet = options("topics").split(",").toSet
-  val kafkaParams: Map[String, String] = options.get("kafkaParams").map { t =>
+  private val topics = options("subscribe").split(",").toSet
+  private val kafkaParams: Map[String, String] = options.get("kafkaParams").map { t =>
     t.split(";").map { s =>
       val a = s.split("->")
       (a(0), a(1))
     }.toMap
   }.getOrElse(Map())
 
-  val K = options.getOrElse("K", "java.lang.String")
-  val V = options.getOrElse("V", "java.lang.String")
-  val KD = options.getOrElse("KD", "kafka.serializer.StringDecoder")
-  val VD = options.getOrElse("VD", "kafka.serializer.StringDecoder")
+  private val preferredHosts = LocationStrategies.PreferConsistent
+  private val startingOffsets = getStartingOffsets
+
+  private def getStartingOffsets = {
+    val clazz = Utils.classForName("org.apache.spark.sql.kafka010.JsonUtils")
+    clazz.getMethod("partitionOffsets", classOf[String])
+        .invoke(null, options("startingOffsets"))
+        .asInstanceOf[Map[TopicPartition, Long]]
```

Comment:
You can store the method to avoid looking up in every call. Also setAccessible will be good to be sure:

```scala
private val partitionOffsetMethod = {
  val clazz = Utils.classForName("org.apache.spark.sql.kafka010.JsonUtils")
  val m = clazz.getMethod("partitionOffsets", classOf[String])
  m.setAccessible(true)
  m
}
```

Reply:
The code is working fine even without explicitly setting accessible to `true` hence not calling `setAccessible`. The other suggestion is incorporated.